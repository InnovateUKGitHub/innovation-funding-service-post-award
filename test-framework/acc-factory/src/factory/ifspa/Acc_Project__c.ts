import { AccOrder } from "../../enum/AccOrder";
import { injectFieldToApex, injectRelationshipToApex } from "../../helpers/apex";
import { AccFieldType, AccRelationshipType } from "../../types/AccFactoryDefinition";
import { AccFactory } from "../AccFactory";
import { competitionBuilder } from "./Competition__c";

const accProjectBuilder = new AccFactory(
  <const>{
    definition: {
      sfdcName: "Acc_Project__c",
      fields: [
        { sfdcName: "Acc_StartDate__c", sfdcType: AccFieldType.DATETIME, nullable: false },
        { sfdcName: "Acc_Duration__c", sfdcType: AccFieldType.NUMBER, nullable: false },
        {
          sfdcName: "Acc_ClaimFrequency__c",
          sfdcType: AccFieldType.SINGLE_PICKLIST,
          values: ["Monthly", "Quarterly"],
          nullable: false,
        },
        { sfdcName: "Acc_ProjectTitle__c", sfdcType: AccFieldType.STRING, nullable: false },
        { sfdcName: "Acc_ProjectNumber__c", sfdcType: AccFieldType.STRING, nullable: false, prefixed: true },
        { sfdcName: "Acc_TSBProjectNumber__c", sfdcType: AccFieldType.NUMBER, nullable: true },
        { sfdcName: "Acc_LegacyID__c", sfdcType: AccFieldType.STRING, nullable: true, prefixed: true },
        { sfdcName: "Acc_WorkdayProjectSetupComplete__c", sfdcType: AccFieldType.CHECKBOX, nullable: false },
        { sfdcName: "Acc_NonFEC__c", sfdcType: AccFieldType.CHECKBOX, nullable: false },
        {
          sfdcName: "Acc_MonitoringLevel__c",
          sfdcType: AccFieldType.SINGLE_PICKLIST,
          values: ["Platinum", "Gold", "Silver", "Bronze", "Internal Assurance"],
          nullable: false,
        },
        {
          sfdcName: "Acc_MonitoringReportSchedule__c",
          sfdcType: AccFieldType.SINGLE_PICKLIST,
          values: ["Monthly", "Quarterly", "6 Monthly", "Yearly", "Internal Assurance"],
          nullable: false,
        },
        {
          sfdcName: "Acc_ProjectStatus__c",
          sfdcType: AccFieldType.SINGLE_PICKLIST,
          values: [
            "Not set",
            "PCL Creation Complete",
            "Offer Letter Sent",
            "Live",
            "On Hold",
            "Final Claim",
            "Closed",
            "Terminated",
          ],
          nullable: false,
        },
        { sfdcName: "Acc_ClaimFrequency__c", sfdcType: AccFieldType.STRING, nullable: false },
        { sfdcName: "Acc_CurrentPeriodNumberHelper__c", sfdcType: AccFieldType.NUMBER, nullable: true },
      ],
      relationships: [
        {
          sfdcName: "Acc_CompetitionId__c",
          sfdcType: AccRelationshipType.SINGLE,
          sffBuilder: competitionBuilder,
          required: true,
        },
      ],
    },
    generator: {
      varName: x => `project${x}`,
    },
  },
  ({ fields, relationships, instanceName, options }) => [
    {
      code: `
Acc_Project__c ${instanceName} = new Acc_Project__c();
${injectRelationshipToApex(instanceName, "Acc_CompetitionId__c", relationships.Acc_CompetitionId__c)}
${injectFieldToApex(options, instanceName, "Acc_StartDate__c", fields.Acc_StartDate__c)}
${injectFieldToApex(options, instanceName, "Acc_Duration__c", fields.Acc_Duration__c)}
${injectFieldToApex(options, instanceName, "Acc_ProjectTitle__c", fields.Acc_ProjectTitle__c)}
${injectFieldToApex(options, instanceName, "Acc_TSBProjectNumber__c", fields.Acc_TSBProjectNumber__c)}
${injectFieldToApex(options, instanceName, "Acc_LegacyID__c", fields.Acc_LegacyID__c)}
${injectFieldToApex(
  options,
  instanceName,
  "Acc_WorkdayProjectSetupComplete__c",
  fields.Acc_WorkdayProjectSetupComplete__c,
)}
insert ${instanceName};
`,
      priority: AccOrder.ACC_PROJECT_LOAD,
    },
    {
      code: `
${injectFieldToApex(options, instanceName, "Acc_ClaimFrequency__c", fields.Acc_ClaimFrequency__c)}
${injectFieldToApex(options, instanceName, "Acc_NonFEC__c", fields.Acc_NonFEC__c)}
${injectFieldToApex(options, instanceName, "Acc_MonitoringLevel__c", fields.Acc_MonitoringLevel__c)}
${injectFieldToApex(options, instanceName, "Acc_MonitoringReportSchedule__c", fields.Acc_MonitoringReportSchedule__c)}
${injectFieldToApex(options, instanceName, "Acc_ProjectStatus__c", fields.Acc_ProjectStatus__c)}
${injectFieldToApex(options, instanceName, "Acc_CurrentPeriodNumberHelper__c", fields.Acc_CurrentPeriodNumberHelper__c)}
upsert ${instanceName};
new Acc_ProjectPeriodProcessor_Batch().start(null);
Acc_ClaimsCreateBatch.start(null);
        `,
      priority: AccOrder.ACC_PROJECT_POSTLOAD,
    },
  ],
);

const currentDate = new Date();

const defaultAccProject = accProjectBuilder.new().set({
  Acc_StartDate__c: new Date(currentDate.getFullYear(), currentDate.getMonth(), 1),
  Acc_Duration__c: 36,
  Acc_ClaimFrequency__c: "Quarterly",
  Acc_ProjectTitle__c: "Title",
  Acc_ProjectNumber__c: "100",
  Acc_LegacyID__c: "100",
  Acc_WorkdayProjectSetupComplete__c: true,
  Acc_NonFEC__c: false,
  Acc_MonitoringLevel__c: "Platinum",
  Acc_MonitoringReportSchedule__c: "Monthly",
  Acc_ProjectStatus__c: "Live",
  Acc_CurrentPeriodNumberHelper__c: 1,
});

export { accProjectBuilder, defaultAccProject };
