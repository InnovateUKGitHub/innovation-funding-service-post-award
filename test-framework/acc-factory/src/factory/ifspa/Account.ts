import { AccOrder } from "../../enum/AccOrder";
import { injectFieldsToApex } from "../../helpers/apex";
import { SffFieldType } from "../../types/SffFactoryDefinition";
import { AccFactory } from "../AccFactory";

const accountBuilder = new AccFactory(
  <const>{
    definition: {
      sfdcName: "Account",
      fields: [
        { sfdcName: "OrgMigrationId__c", sfdcType: SffFieldType.STRING, nullable: false, prefixed: true },
        { sfdcName: "Name", sfdcType: SffFieldType.STRING, nullable: false },
        { sfdcName: "BillingStreet", sfdcType: SffFieldType.STRING, nullable: false },
        { sfdcName: "BillingCity", sfdcType: SffFieldType.STRING, nullable: false },
        { sfdcName: "BillingState", sfdcType: SffFieldType.STRING, nullable: false },
        { sfdcName: "BillingPostalCode", sfdcType: SffFieldType.STRING, nullable: false },
        { sfdcName: "BillingCountry", sfdcType: SffFieldType.STRING, nullable: false },
      ],
      relationships: [],
    },
    generator: {
      varName: x => `account${x}`,
    },
  },
  ({ fields, instanceName, options }) => [
    {
      code: `
Account ${instanceName} = new Account();
${injectFieldsToApex(options, instanceName, fields)}
insert ${instanceName};
      `,
      priority: AccOrder.ACCOUNT_LOAD,
    },
  ],
);

const defaultAccount = accountBuilder
  .new()
  .setField("BillingStreet", "North Star Avenue")
  .setField("BillingCity", "Swindon")
  .setField("BillingState", "Wiltshire")
  .setField("BillingPostalCode", "SN2 1SZ")
  .setField("BillingCountry", "United Kingdom");

export { accountBuilder, defaultAccount };
