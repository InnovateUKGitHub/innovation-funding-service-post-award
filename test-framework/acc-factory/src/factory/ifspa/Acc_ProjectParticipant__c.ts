import { AccOrder } from "../../enum/AccOrder";
import { injectFieldsToApex, injectRelationshipToApex } from "../../helpers/apex";
import { AccFieldType, AccRelationshipType } from "../../types/AccFactoryDefinition";
import { AccFactory } from "../AccFactory";
import { accProjectBuilder } from "./Acc_Project__c";
import { accountBuilder } from "./Account";

const accProjectParticipantBuilder = new AccFactory(
  <const>{
    definition: {
      sfdcName: "Acc_ProjectParticipant__c",
      fields: [
        { sfdcName: "ParticipantMigrationID__c", sfdcType: AccFieldType.STRING, nullable: false, prefixed: true },
        {
          sfdcName: "Acc_ParticipantType__c",
          sfdcType: AccFieldType.SINGLE_PICKLIST,
          values: [
            "Business",
            "Knowledge base",
            "Research",
            "Research and Technology Organisation (RTO)",
            "Public sector, charity or non Je-S registered research organisation",
          ],
          nullable: false,
        },
        {
          sfdcName: "Acc_ParticipantSize__c",
          sfdcType: AccFieldType.SINGLE_PICKLIST,
          values: ["Small", "Medium", "Large", "Academic", "Unknown", "Micro", "Small1"],
          nullable: false,
        },
        {
          sfdcName: "Acc_ProjectRole__c",
          sfdcType: AccFieldType.SINGLE_PICKLIST,
          values: ["Collaborator", "Lead"],
          nullable: false,
        },
        {
          sfdcName: "Acc_AuditReportFrequency__c",
          sfdcType: AccFieldType.SINGLE_PICKLIST,
          values: [
            "Never, for this project",
            "With all claims",
            "With the first and last claim only",
            "With the last claim only",
            "With the first claim, last claim and on every anniversary of the project start date",
            "Quarterly",
          ],
          nullable: false,
        },
        {
          sfdcName: "Acc_ParticipantStatus__c",
          sfdcType: AccFieldType.SINGLE_PICKLIST,
          values: [
            "Pending",
            "Active",
            "On Hold",
            "Invuluntary Withdrawal",
            "Voluntary Withdrawal",
            "Migrated - Withdrawn",
            "Closed",
          ],
          nullable: false,
        },
        { sfdcName: "Acc_Award_Rate__c", sfdcType: AccFieldType.NUMBER, nullable: false },
        { sfdcName: "Acc_Cap_Limit__c", sfdcType: AccFieldType.NUMBER, nullable: false },
        { sfdcName: "Acc_FlaggedParticipant__c", sfdcType: AccFieldType.CHECKBOX, nullable: false },
        { sfdcName: "Acc_OverheadRate__c", sfdcType: AccFieldType.NUMBER, nullable: false },
        {
          sfdcName: "Acc_ParticipantProjectReportingType__c",
          sfdcType: AccFieldType.SINGLE_PICKLIST,
          values: ["Public", "Private"],
          nullable: false,
        },
        {
          sfdcName: "Acc_OrganisationType__c",
          sfdcType: AccFieldType.SINGLE_PICKLIST,
          values: ["Industrial", "Academic"],
          nullable: false,
        },
        { sfdcName: "Acc_CreateProfiles__c", sfdcType: AccFieldType.CHECKBOX, nullable: false },
        { sfdcName: "Acc_CreateClaims__c", sfdcType: AccFieldType.CHECKBOX, nullable: false },
      ],
      relationships: [
        {
          sfdcName: "Acc_ProjectId__c",
          sfdcType: AccRelationshipType.SINGLE,
          sffBuilder: accProjectBuilder,
          required: true,
        },
        {
          sfdcName: "Acc_AccountId__c",
          sfdcType: AccRelationshipType.SINGLE,
          sffBuilder: accountBuilder,
          required: true,
        },
      ],
    },
    generator: {
      varName: x => `projectParticipant${x}`,
    },
  },
  ({ fields, relationships, instanceName, options }) => [
    {
      code: `
Acc_ProjectParticipant__c ${instanceName} = new Acc_ProjectParticipant__c();
${injectFieldsToApex(options, instanceName, fields)}
${injectRelationshipToApex(instanceName, "Acc_ProjectId__c", relationships.Acc_ProjectId__c)}
${injectRelationshipToApex(instanceName, "Acc_AccountId__c", relationships.Acc_AccountId__c)}
insert ${instanceName};
      `,
      priority: AccOrder.ACC_PROJECT_PARTICIPANT_LOAD,
    },
  ],
);

const defaultAccProjectParticipant = accProjectParticipantBuilder
  .new()
  .set({
    ParticipantMigrationID__c: "004001",
    Acc_ParticipantType__c: "Business",
    Acc_ParticipantSize__c: "Medium",
    Acc_ProjectRole__c: "Lead",
    Acc_AuditReportFrequency__c: "With all claims",
    Acc_ParticipantStatus__c: "Active",
    Acc_Award_Rate__c: 50,
    Acc_Cap_Limit__c: 50,
    Acc_FlaggedParticipant__c: false,
    Acc_OverheadRate__c: 20,
    Acc_ParticipantProjectReportingType__c: "Public",
    Acc_OrganisationType__c: "Industrial",
    Acc_CreateProfiles__c: false,
    Acc_CreateClaims__c: false,
  })
  .copy();

export { accProjectParticipantBuilder, defaultAccProjectParticipant };
