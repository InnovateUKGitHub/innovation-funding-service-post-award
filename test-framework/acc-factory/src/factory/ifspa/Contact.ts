import { AccOrder } from "../../enum/AccOrder";
import { injectFieldsToApex, injectRelationshipToApex } from "../../helpers/apex";
import { AccFieldType, AccRelationshipType } from "../../types/AccFactoryDefinition";
import { AccFactory } from "../AccFactory";
import { accountBuilder } from "./Account";

const contactBuilder = new AccFactory(
  <const>{
    definition: {
      sfdcName: "Contact",
      fields: [
        { sfdcName: "ContactMigrationId__c", sfdcType: AccFieldType.STRING, nullable: false, prefixed: true },
        { sfdcName: "FirstName", sfdcType: AccFieldType.STRING, nullable: false },
        { sfdcName: "LastName", sfdcType: AccFieldType.STRING, nullable: false },
        { sfdcName: "Email", sfdcType: AccFieldType.STRING, nullable: false, prefixed: true },
      ],
      relationships: [
        { sfdcName: "AccountId", sfdcType: AccRelationshipType.SINGLE, sffBuilder: accountBuilder, required: true },
      ],
    },
    generator: {
      varName: x => `contact${x}`,
    },
  },
  ({ fields, relationships, instanceName, options }) => [
    {
      code: `
Contact ${instanceName} = new Contact();
${injectFieldsToApex(options, instanceName, fields)}
${injectRelationshipToApex(instanceName, "AccountId", relationships.AccountId)}
insert ${instanceName};
      `,
      priority: AccOrder.CONTACT_LOAD,
    },
  ],
);

export { contactBuilder };
