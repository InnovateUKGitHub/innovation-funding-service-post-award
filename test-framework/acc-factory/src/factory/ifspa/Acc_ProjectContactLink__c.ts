import { AccOrder } from "../../enum/AccOrder";
import { injectFieldsToApex, injectRelationshipToApex } from "../../helpers/apex";
import { AccFieldType, AccRelationshipType } from "../../types/AccFactoryDefinition";
import { AccFactory } from "../AccFactory";
import { accProjectBuilder } from "./Acc_Project__c";
import { accountBuilder } from "./Account";
import { contactBuilder } from "./Contact";
import { userBuilder } from "./User";

const accProjectContactLinkBuilder = new AccFactory(
  <const>{
    definition: {
      sfdcName: "Acc_ProjectContactLink__c",
      fields: [
        {
          sfdcName: "Acc_Role__c",
          sfdcType: AccFieldType.SINGLE_PICKLIST,
          values: [
            "Finance contact",
            "Monitoring officer",
            "Project Manager",
            "Innovation Lead",
            "IPM",
            "Claims Notification",
            "PCR Notification",
            "Investor",
            "Associate",
            "Company Supervisor",
            "Main Company Contact",
            "KB Supervisor",
            "Main KB Contact",
            "KB Admin",
            "KB Finance",
            "Final Report Assessor",
            "Contracting Authority",
            "Relationship Manager",
          ],
          nullable: false,
        },
        { sfdcName: "Acc_EmailOfSFContact__c", sfdcType: AccFieldType.STRING, nullable: false, prefixed: true },
      ],
      relationships: [
        {
          sfdcName: "Acc_ProjectId__c",
          sfdcType: AccRelationshipType.SINGLE,
          sffBuilder: accProjectBuilder,
          required: true,
        },
        {
          sfdcName: "Acc_AccountId__c",
          sfdcType: AccRelationshipType.SINGLE,
          sffBuilder: accountBuilder,
          required: true,
        },
        {
          sfdcName: "Acc_ContactId__c",
          sfdcType: AccRelationshipType.SINGLE,
          sffBuilder: contactBuilder,
          required: true,
        },
        { sfdcName: "Acc_UserId__c", sfdcType: AccRelationshipType.SINGLE, sffBuilder: userBuilder, required: false },
      ],
    },
    generator: {
      varName: x => `pcl${x}`,
    },
  },
  ({ fields, relationships, instanceName, options }) => [
    {
      code: `
Acc_ProjectContactLink__c ${instanceName} = new Acc_ProjectContactLink__c();
${injectFieldsToApex(options, instanceName, fields)}
${injectRelationshipToApex(instanceName, "Acc_ProjectId__c", relationships.Acc_ProjectId__c)}
${injectRelationshipToApex(instanceName, "Acc_AccountId__c", relationships.Acc_AccountId__c)}
${injectRelationshipToApex(instanceName, "Acc_ContactId__c", relationships.Acc_ContactId__c)}
${injectRelationshipToApex(instanceName, "Acc_UserId__c", relationships.Acc_UserId__c)}
insert ${instanceName};
      `,
      priority: AccOrder.ACC_PROJECT_CONTACT_LINK_LOAD,
    },
  ],
);

export { accProjectContactLinkBuilder };
