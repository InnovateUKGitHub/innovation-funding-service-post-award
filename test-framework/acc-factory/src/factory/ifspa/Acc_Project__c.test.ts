import { describe, expect, test } from "@jest/globals";
import { competitionBuilder } from "./Competition__c";
import { accProjectBuilder } from "./Acc_Project__c";
import { buildApex } from "../../helpers/apex";
import { accountBuilder } from "./Account";
import { userBuilder } from "./User";
import { contactBuilder } from "./Contact";
import { accProjectContactLinkBuilder } from "./Acc_ProjectContactLink__c";
import { accProjectParticipantBuilder } from "./Acc_ProjectParticipant__c";

const competition = competitionBuilder
  .new()
  .setField("Acc_CompetitionCode__c", "100")
  .setField("Acc_CompetitionName__c", "Competition")
  .setField("Acc_CompetitionType__c", "KTP");

const account = accountBuilder
  .new()
  .setField("OrgMigrationId__c", "001")
  .setField("Name", "xgov")
  .setField("BillingStreet", "North Star Avenue")
  .setField("BillingCity", "Swindon")
  .setField("BillingState", "Wiltshire")
  .setField("BillingPostalCode", "SN2 1SZ")
  .setField("BillingCountry", "United Kingdom");

const project = accProjectBuilder
  .new()
  .setRelationship("Acc_CompetitionId__c", competition)
  .setField("Acc_StartDate__c", new Date("2024-01-16 12:59:42"))
  .setField("Acc_Duration__c", 36)
  .setField("Acc_ClaimFrequency__c", "Quarterly")
  .setField("Acc_ProjectTitle__c", "Title")
  .setField("Acc_ProjectNumber__c", "100")
  .setField("Acc_LegacyID__c", "100")
  .setField("Acc_WorkdayProjectSetupComplete__c", true)
  .setField("Acc_NonFEC__c", false)
  .setField("Acc_MonitoringLevel__c", "Platinum")
  .setField("Acc_MonitoringReportSchedule__c", "Monthly")
  .setField("Acc_ProjectStatus__c", "Live");

const contact = contactBuilder
  .new()
  .setRelationship("AccountId", account)
  .setField("ContactMigrationId__c", "001")
  .setField("Email", "austria@x.gov.uk")
  .setField("FirstName", "Austria")
  .setField("LastName", "Hedges");

const user = userBuilder
  .new()
  .setRelationship("ContactId", contact)
  .setField("Username", "austria@x.gov.uk")
  .setField("Email", "austria@x.gov.uk")
  .setField("FirstName", "Austria")
  .setField("LastName", "Hedges")
  .setField("Alias", "xgovuk")
  .setField("CommunityNickname", "austria")
  .setField("EmailEncodingKey", "UTF-8")
  .setField("LocaleSidKey", "en_GB")
  .setField("LanguageLocaleKey", "en_US")
  .setField("TimeZoneSidKey", "Europe/London")
  .setField("ProfileId", "00e58000001ITpLAAW");

const projectContactLink = accProjectContactLinkBuilder
  .new()
  .setRelationship("Acc_AccountId__c", account)
  .setRelationship("Acc_ContactId__c", contact)
  .setRelationship("Acc_ProjectId__c", project)
  .setField("Acc_EmailOfSFContact__c", "noemail@noemail.com")
  .setField("Acc_Role__c", "Monitoring officer");

const projectParticipant = accProjectParticipantBuilder
  .new()
  .setRelationship("Acc_AccountId__c", account)
  .setRelationship("Acc_ProjectId__c", project)
  .setField("ParticipantMigrationID__c", "004001")
  .setField("Acc_ParticipantType__c", "Business")
  .setField("Acc_ParticipantSize__c", "Medium")
  .setField("Acc_ProjectRole__c", "Lead")
  .setField("Acc_AuditReportFrequency__c", "With all claims")
  .setField("Acc_ParticipantStatus__c", "Active")
  .setField("Acc_Award_Rate__c", 50)
  .setField("Acc_Cap_Limit__c", 50)
  .setField("Acc_FlaggedParticipant__c", false)
  .setField("Acc_OverheadRate__c", 20)
  .setField("Acc_ParticipantProjectReportingType__c", "Public")
  .setField("Acc_OrganisationType__c", "Industrial")
  .setField("Acc_CreateProfiles__c", false)
  .setField("Acc_CreateClaims__c", false);

describe("Project Builder", () => {
  test("Expect project to be built, linking correctly to competition", () => {
    const code = buildApex({ instances: [project, competition] });
    expect(code).toMatchSnapshot();
  });

  test("Expect everything to be harmonious", () => {
    const code = buildApex({
      instances: [project, competition, account, contact, user, projectContactLink, projectParticipant],
    });
    expect(code).toMatchSnapshot();
  });

  test("Expect everything to be harmonious with a prefix", () => {
    const code = buildApex({
      instances: [project, competition, account, contact, user, projectContactLink, projectParticipant],
      options: { prefix: "hello." },
    });
    expect(code).toMatchSnapshot();
  });
});
