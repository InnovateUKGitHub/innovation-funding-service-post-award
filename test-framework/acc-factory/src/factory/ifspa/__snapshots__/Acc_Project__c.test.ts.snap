// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Project Builder Expect everything to be harmonious 1`] = `
"
Account account1 = new Account();
account1.OrgMigrationId__c = '001';
account1.Name = 'xgov';
account1.BillingStreet = 'North Star Avenue';
account1.BillingCity = 'Swindon';
account1.BillingState = 'Wiltshire';
account1.BillingPostalCode = 'SN2 1SZ';
account1.BillingCountry = 'United Kingdom';
insert account1;
      
Competition__c competition0 = new Competition__c();
competition0.Acc_CompetitionCode__c = '100';
competition0.Acc_CompetitionType__c = 'KTP';
competition0.Acc_CompetitionName__c = 'Competition';
insert competition0;
      
Contact contact0 = new Contact();
contact0.ContactMigrationId__c = '001';
contact0.FirstName = 'Austria';
contact0.LastName = 'Hedges';
contact0.Email = 'austria@x.gov.uk';
contact0.AccountId = account1.Id;
insert contact0;
      
User user1 = new User();
user1.Username = 'austria@x.gov.uk';
user1.Email = 'austria@x.gov.uk';
user1.FirstName = 'Austria';
user1.LastName = 'Hedges';
user1.Alias = 'xgovuk';
user1.CommunityNickname = 'austria';
user1.EmailEncodingKey = 'UTF-8';
user1.LocaleSidKey = 'en_GB';
user1.LanguageLocaleKey = 'en_US';
user1.TimeZoneSidKey = 'Europe/London';
user1.ProfileId = '00e58000001ITpLAAW';
user1.ContactId = contact0.Id;
insert user1;
      
Acc_Project__c project1 = new Acc_Project__c();
project1.Acc_CompetitionId__c = competition0.Id;
project1.Acc_StartDate__c = date.valueOf('2024-01-16 12:59:42');
project1.Acc_Duration__c = 36;
project1.Acc_ProjectTitle__c = 'Title';
// project1.Acc_TSBProjectNumber__c field is not defined
project1.Acc_LegacyID__c = '100';
project1.Acc_WorkdayProjectSetupComplete__c = true;
insert project1;

Acc_ProjectParticipant__c projectParticipant2 = new Acc_ProjectParticipant__c();
projectParticipant2.ParticipantMigrationID__c = '004001';
projectParticipant2.Acc_ParticipantType__c = 'Business';
projectParticipant2.Acc_ParticipantSize__c = 'Medium';
projectParticipant2.Acc_ProjectRole__c = 'Lead';
projectParticipant2.Acc_AuditReportFrequency__c = 'With all claims';
projectParticipant2.Acc_ParticipantStatus__c = 'Active';
projectParticipant2.Acc_Award_Rate__c = 50;
projectParticipant2.Acc_Cap_Limit__c = 50;
projectParticipant2.Acc_FlaggedParticipant__c = false;
projectParticipant2.Acc_OverheadRate__c = 20;
projectParticipant2.Acc_ParticipantProjectReportingType__c = 'Public';
projectParticipant2.Acc_OrganisationType__c = 'Industrial';
projectParticipant2.Acc_CreateProfiles__c = false;
projectParticipant2.Acc_CreateClaims__c = false;
projectParticipant2.Acc_ProjectId__c = project1.Id;
projectParticipant2.Acc_AccountId__c = account1.Id;
insert projectParticipant2;
      
Acc_ProjectContactLink__c pcl0 = new Acc_ProjectContactLink__c();
pcl0.Acc_Role__c = 'Monitoring officer';
pcl0.Acc_EmailOfSFContact__c = 'noemail@noemail.com';
pcl0.Acc_ProjectId__c = project1.Id;
pcl0.Acc_AccountId__c = account1.Id;
pcl0.Acc_ContactId__c = contact0.Id;
// pcl0.Acc_UserId__c relationship is not defined
insert pcl0;
      
project1.Acc_ClaimFrequency__c = 'Quarterly';
project1.Acc_NonFEC__c = false;
project1.Acc_MonitoringLevel__c = 'Platinum';
project1.Acc_MonitoringReportSchedule__c = 'Monthly';
project1.Acc_ProjectStatus__c = 'Live';
// project1.Acc_CurrentPeriodNumberHelper__c field is not defined
upsert project1;
new Acc_ProjectPeriodProcessor_Batch().start(null);
Acc_ClaimsCreateBatch.start(null);
        "
`;

exports[`Project Builder Expect everything to be harmonious with a prefix 1`] = `
"
Account account1 = new Account();
account1.OrgMigrationId__c = 'hello.001';
account1.Name = 'xgov';
account1.BillingStreet = 'North Star Avenue';
account1.BillingCity = 'Swindon';
account1.BillingState = 'Wiltshire';
account1.BillingPostalCode = 'SN2 1SZ';
account1.BillingCountry = 'United Kingdom';
insert account1;
      
Competition__c competition0 = new Competition__c();
competition0.Acc_CompetitionCode__c = 'hello.100';
competition0.Acc_CompetitionType__c = 'KTP';
competition0.Acc_CompetitionName__c = 'Competition';
insert competition0;
      
Contact contact0 = new Contact();
contact0.ContactMigrationId__c = 'hello.001';
contact0.FirstName = 'Austria';
contact0.LastName = 'Hedges';
contact0.Email = 'hello.austria@x.gov.uk';
contact0.AccountId = account1.Id;
insert contact0;
      
User user1 = new User();
user1.Username = 'hello.austria@x.gov.uk';
user1.Email = 'hello.austria@x.gov.uk';
user1.FirstName = 'Austria';
user1.LastName = 'Hedges';
user1.Alias = 'xgovuk';
user1.CommunityNickname = 'hello.austria';
user1.EmailEncodingKey = 'UTF-8';
user1.LocaleSidKey = 'en_GB';
user1.LanguageLocaleKey = 'en_US';
user1.TimeZoneSidKey = 'Europe/London';
user1.ProfileId = '00e58000001ITpLAAW';
user1.ContactId = contact0.Id;
insert user1;
      
Acc_Project__c project1 = new Acc_Project__c();
project1.Acc_CompetitionId__c = competition0.Id;
project1.Acc_StartDate__c = date.valueOf('2024-01-16 12:59:42');
project1.Acc_Duration__c = 36;
project1.Acc_ProjectTitle__c = 'Title';
// project1.Acc_TSBProjectNumber__c field is not defined
project1.Acc_LegacyID__c = 'hello.100';
project1.Acc_WorkdayProjectSetupComplete__c = true;
insert project1;

Acc_ProjectParticipant__c projectParticipant2 = new Acc_ProjectParticipant__c();
projectParticipant2.ParticipantMigrationID__c = 'hello.004001';
projectParticipant2.Acc_ParticipantType__c = 'Business';
projectParticipant2.Acc_ParticipantSize__c = 'Medium';
projectParticipant2.Acc_ProjectRole__c = 'Lead';
projectParticipant2.Acc_AuditReportFrequency__c = 'With all claims';
projectParticipant2.Acc_ParticipantStatus__c = 'Active';
projectParticipant2.Acc_Award_Rate__c = 50;
projectParticipant2.Acc_Cap_Limit__c = 50;
projectParticipant2.Acc_FlaggedParticipant__c = false;
projectParticipant2.Acc_OverheadRate__c = 20;
projectParticipant2.Acc_ParticipantProjectReportingType__c = 'Public';
projectParticipant2.Acc_OrganisationType__c = 'Industrial';
projectParticipant2.Acc_CreateProfiles__c = false;
projectParticipant2.Acc_CreateClaims__c = false;
projectParticipant2.Acc_ProjectId__c = project1.Id;
projectParticipant2.Acc_AccountId__c = account1.Id;
insert projectParticipant2;
      
Acc_ProjectContactLink__c pcl0 = new Acc_ProjectContactLink__c();
pcl0.Acc_Role__c = 'Monitoring officer';
pcl0.Acc_EmailOfSFContact__c = 'hello.noemail@noemail.com';
pcl0.Acc_ProjectId__c = project1.Id;
pcl0.Acc_AccountId__c = account1.Id;
pcl0.Acc_ContactId__c = contact0.Id;
// pcl0.Acc_UserId__c relationship is not defined
insert pcl0;
      
project1.Acc_ClaimFrequency__c = 'Quarterly';
project1.Acc_NonFEC__c = false;
project1.Acc_MonitoringLevel__c = 'Platinum';
project1.Acc_MonitoringReportSchedule__c = 'Monthly';
project1.Acc_ProjectStatus__c = 'Live';
// project1.Acc_CurrentPeriodNumberHelper__c field is not defined
upsert project1;
new Acc_ProjectPeriodProcessor_Batch().start(null);
Acc_ClaimsCreateBatch.start(null);
        "
`;

exports[`Project Builder Expect project to be built, linking correctly to competition 1`] = `
"
Competition__c competition0 = new Competition__c();
competition0.Acc_CompetitionCode__c = '100';
competition0.Acc_CompetitionType__c = 'KTP';
competition0.Acc_CompetitionName__c = 'Competition';
insert competition0;
      
Acc_Project__c project1 = new Acc_Project__c();
project1.Acc_CompetitionId__c = competition0.Id;
project1.Acc_StartDate__c = date.valueOf('2024-01-16 12:59:42');
project1.Acc_Duration__c = 36;
project1.Acc_ProjectTitle__c = 'Title';
// project1.Acc_TSBProjectNumber__c field is not defined
project1.Acc_LegacyID__c = '100';
project1.Acc_WorkdayProjectSetupComplete__c = true;
insert project1;

project1.Acc_ClaimFrequency__c = 'Quarterly';
project1.Acc_NonFEC__c = false;
project1.Acc_MonitoringLevel__c = 'Platinum';
project1.Acc_MonitoringReportSchedule__c = 'Monthly';
project1.Acc_ProjectStatus__c = 'Live';
// project1.Acc_CurrentPeriodNumberHelper__c field is not defined
upsert project1;
new Acc_ProjectPeriodProcessor_Batch().start(null);
Acc_ClaimsCreateBatch.start(null);
        "
`;
